File: python/src/translated/Soldier.py
import random
import funcsplayer0.Util #TODO Idk why this is here in the java version
from battlecode25.stubs import *

class Soldier(Robot):
    turn_count = 0
    rng = random.Random(6147)

    target = None
    original_location = None
    nearest_tower = None
    prev_target = None

    paint_tower_pattern = None
    money_tower_pattern = None
    defense_tower_pattern = None

    painting_ruin_loc = None
    painting_tower_type = None
    painting_turns = 0
    turns_without_attack = 0

    cur_resource = None
    override = False

    has_enemy_paint = False
    need_mopper_at = None

    # Array containing all the possible movement directions
    directions = [
        Direction.NORTH,
        Direction.NORTHEAST,
        Direction.EAST,
        Direction.SOUTHEAST,
        Direction.SOUTH,
        Direction.SOUTHWEST,
        Direction.WEST,
        Direction.NORTHWEST,
    ]

    def __init__(self, rc: RobotController):
        super().__init__(rc)
        self.rc = rc
        # Initialize the map recorder with map dimensions
        MapRecorder.init_map(rc.getMapWidth(), rc.getMapHeight())
        self.paint_tower_pattern = rc.getTowerPattern(UnitType.LEVEL_ONE_PAINT_TOWER)
        self.money_tower_pattern = rc.getTowerPattern(UnitType.LEVEL_ONE_MONEY_TOWER)
        self.defense_tower_pattern = rc.getTowerPattern(UnitType.LEVEL_ONE_DEFENSE_TOWER)

    def play(self):
        translated.Util.init(self.rc)
        translated.Explore.init(self.rc)
        translated.Pathfinding.init(self.rc)
        translated.Pathfinding.initTurn()
        self.target = translated.Explore.getExploreTarget()

        while True:
            nearby_tiles = self.rc.senseNearbyMapInfos()
            cur_ruin = None
            ruin_dist = 999999
            tower_dist = 999999

            for tile in nearby_tiles:
                potential_tower = self.rc.senseRobotAtLocation(tile.getMapLocation())
                if tile.hasRuin() and potential_tower is None:
                    dist = tile.getMapLocation().distanceSquaredTo(self.rc.getLocation())
                    if dist < ruin_dist:
                        cur_ruin = tile
                        ruin_dist = dist
                elif tile.hasRuin() and potential_tower is not None:
                    if potential_tower.getTeam() == self.rc.getTeam():
                        dist = tile.getMapLocation().distanceSquaredTo(self.rc.getLocation())
                        if dist < tower_dist:
                            self.nearest_tower = tile.getMapLocation()
                            tower_dist = dist
                            if self.rc.getPaint() < 100:
                                if self.rc.canTransferPaint(self.nearest_tower, -75):
                                    self.rc.transferPaint(self.nearest_tower, -75)
                        if self.rc.canUpgradeTower(tile.getMapLocation()):
                            self.rc.upgradeTower(tile.getMapLocation())
                            print("Tower was upgraded!")

            if cur_ruin is not None:
                self.painting_ruin_loc = cur_ruin.getMapLocation()
                self.painting_tower_type = self.get_new_tower_type(self.rc)
                self.turns_without_attack = 0
                self.painting_turns = 0
                self.run_paint_pattern(self.rc)
                return

            if self.rc.getMovementCooldownTurns() > 10:
                Clock.yield()

            if self.rc.getLocation().x % 4 == 0 and self.rc.getLocation().y % 4 == 3:
                has_mark = False
                for pattern_tile in self.rc.senseNearbyMapInfos(self.rc.getLocation(), -1):
                    if pattern_tile.getMark() != pattern_tile.getPaint() and pattern_tile.getMark() != PaintType.EMPTY:
                        if pattern_tile.getMark() != PaintType.EMPTY:
                            has_mark = True
                        if self.rc.canAttack(pattern_tile.getMapLocation()):
                            self.rc.attack(pattern_tile.getMapLocation(), pattern_tile.getMark() == PaintType.ALLY_SECONDARY)

                if self.rc.getNumberTowers() > 8 and not has_mark and self.rc.canMarkResourcePattern(self.rc.getLocation()):
                    self.cur_resource = self.rc.getLocation()
                    self.target = self.cur_resource
                    self.rc.markResourcePattern(self.cur_resource)

            if self.cur_resource is not None:
                direction = self.rc.getLocation().directionTo(self.cur_resource)
                if self.rc.getMovementCooldownTurns() <= 10:
                    if self.rc.canMove(direction):
                        self.rc.move(direction)
                    elif self.rc.canMove(direction.rotateRight()):
                        self.rc.move(direction.rotateRight())
                    if self.rc.canCompleteResourcePattern(self.cur_resource):
                        self.rc.completeResourcePattern(self.cur_resource)
                        self.cur_resource = None
                        break
                    for pattern_tile in self.rc.senseNearbyMapInfos(self.cur_resource, -1):
                        if pattern_tile.getMark() != pattern_tile.getPaint() and pattern_tile.getMark() != PaintType.EMPTY:
                            if self.rc.canAttack(pattern_tile.getMapLocation()):
                                self.rc.attack(pattern_tile.getMapLocation(), pattern_tile.getMark() == PaintType.ALLY_SECONDARY)

            if cur_ruin is None and self.cur_resource is None:
                try:
                    if translated.Util.distance(self.rc.getLocation(), self.target) <= 5 or not self.rc.isMovementReady():
                        old_target = self.target
                        while translated.Util.distance(translated.Explore.getExploreTarget(), old_target) <= 20:
                            translated.Explore.getNewTarget(10)
                        self.target = translated.Explore.explore_target

                    self.rc.setIndicatorString(f"Moving toward target at {self.target}")
                    translated.Pathfinding.move(self.target, False)
                except Exception as e:
                    print(f"{self.rc.getType()} Exception")

            Clock.yield()

    @staticmethod
    def is_within_pattern(ruin_loc: MapLocation, paint_loc: MapLocation) -> bool:
        return abs(paint_loc.x - ruin_loc.x) <= 2 and abs(paint_loc.y - ruin_loc.y) <= 2 and not ruin_loc.equals(paint_loc)

    @staticmethod
    def get_is_secondary(ruin_loc: MapLocation, paint_loc: MapLocation, tower_type: UnitType) -> bool:
        if not Soldier.is_within_pattern(ruin_loc, paint_loc):
            return False
        col = paint_loc.x - ruin_loc.x + 2
        row = paint_loc.y - ruin_loc.y + 2
        if tower_type == UnitType.LEVEL_ONE_PAINT_TOWER:
            return Soldier.paint_tower_pattern[row][col]
        elif tower_type == UnitType.LEVEL_ONE_MONEY_TOWER:
            return Soldier.money_tower_pattern[row][col]
        else:
            return Soldier.defense_tower_pattern[row][col]

    @staticmethod
    def run_paint_pattern(rc: RobotController) -> None:
        rc.setIndicatorString("SEARCHING FOR NEARBY ROBOTS")
        for robot in rc.senseNearbyRobots(-1):
            if robot.team == rc.getTeam() and robot.type == UnitType.SOLDIER and \
                    Soldier.painting_ruin_loc.distanceSquaredTo(robot.location) < rc.getLocation().distanceSquaredTo(Soldier.painting_ruin_loc):
                translated.Pathfinding.move(Soldier.target, False)
                return

        # paint tiles for the pattern
        if rc.isActionReady():
            infos = rc.senseNearbyMapInfos()
            attacked = False
            for info in infos:
                paint_loc = info.getMapLocation()
                should_be_secondary = Soldier.get_is_secondary(Soldier.painting_ruin_loc, paint_loc, Soldier.painting_tower_type)
                if rc.canAttack(paint_loc) and (info.getPaint() == PaintType.EMPTY or info.getPaint().isSecondary() != should_be_secondary) \
                        and Soldier.is_within_pattern(Soldier.painting_ruin_loc, paint_loc):
                    rc.attack(paint_loc, should_be_secondary)
                    attacked = True
                    Soldier.turns_without_attack = 0
                    break
            if not attacked:
                Soldier.turns_without_attack += 1

        direction = rc.getLocation().directionTo(Soldier.painting_ruin_loc)

        if rc.getMovementCooldownTurns() > 10:
            # Do nothing
            pass
        elif rc.canMove(direction):
            rc.move(direction)
        elif rc.canMove(direction.rotateRight()):
            rc.move(direction.rotateRight())

        if rc.canCompleteTowerPattern(Soldier.painting_tower_type, Soldier.painting_ruin_loc):
            rc.completeTowerPattern(Soldier.painting_tower_type, Soldier.painting_ruin_loc)

    @staticmethod
    def get_new_tower_type(rc: RobotController) -> UnitType:
        random_number = random.randint(0, 6)
        in_middle_x = 0.3 * rc.getMapWidth() < rc.getLocation().x < 0.7 * rc.getMapWidth()
        in_middle_y = 0.3 * rc.getMapHeight() < rc.getLocation().y < 0.7 * rc.getMapHeight()

        if in_middle_x and in_middle_y:
            return UnitType.LEVEL_ONE_DEFENSE_TOWER
        elif rc.getNumberTowers() < 4:
            return UnitType.LEVEL_ONE_MONEY_TOWER
        elif random_number in {0, 1, 2}:
            return UnitType.LEVEL_ONE_PAINT_TOWER
        else:
            return UnitType.LEVEL_ONE_MONEY_TOWER

File: java/src/funcsplayer2/Soldier.java
package funcsplayer2;

import java.util.Random;

import battlecode.common.Clock;
import battlecode.common.Direction;
import battlecode.common.GameActionException;
import battlecode.common.MapInfo;
import battlecode.common.MapLocation;
import battlecode.common.PaintType;
import battlecode.common.RobotController;
import battlecode.common.RobotInfo;
import battlecode.common.UnitType;
import funcsplayer0.Util;

public class Soldier extends Robot {
	static int turnCount = 0;

    static final Random rng = new Random(6147);
    static MapLocation target;
    static MapLocation originalLocation = null;
    static MapLocation nearestTower;
    static MapLocation prevTarget = null;
    
    static boolean[][] paintTowerPattern = null;
    static boolean[][] moneyTowerPattern = null;
    static boolean[][] defenseTowerPattern = null;

    static MapLocation paintingRuinLoc = null;
    static UnitType paintingTowerType = null;
    static int paintingTurns = 0;
    static int turnsWithoutAttack = 0;
    
    MapLocation curResource = null;
    boolean override = false;

    static boolean hasEnemyPaint = false;
    static MapLocation needMopperAt;
    
    static boolean weaving = false;
	static boolean moveAway = false;
	static MapLocation myEnemyTower = null;

    /** Array containing all the possible movement directions. */
    static final Direction[] directions = {
        Direction.NORTH,
        Direction.NORTHEAST,
        Direction.EAST,
        Direction.SOUTHEAST,
        Direction.SOUTH,
        Direction.SOUTHWEST,
        Direction.WEST,
        Direction.NORTHWEST,
    };

    Soldier(RobotController rc) throws GameActionException {
        super(rc);
        Robot.rc = rc;
        // Initialization that needs to happen when a Soldier instance is created
        MapRecorder.initMap(rc.getMapWidth(), rc.getMapHeight()); // Initialize the map recorder with map dimensions
        paintTowerPattern = rc.getTowerPattern(UnitType.LEVEL_ONE_PAINT_TOWER);
        moneyTowerPattern = rc.getTowerPattern(UnitType.LEVEL_ONE_MONEY_TOWER);
        defenseTowerPattern = rc.getTowerPattern(UnitType.LEVEL_ONE_DEFENSE_TOWER);
    }

    @Override
	void play() throws GameActionException {
    	funcsplayer2.Util.init(rc);
    	funcsplayer2.Explore.init(rc);
    	funcsplayer2.BugNav.init(rc);
        funcsplayer2.Pathfinding.init(rc);
        funcsplayer2.Pathfinding.initTurn();
        target = funcsplayer2.Explore.getExploreTarget();
    	while (true) {
    		
    		if (rc.getPaint() < 100) {
    			anotherfuncsplayer.Pathfinding.move(nearestTower, false);
				if (rc.canTransferPaint(nearestTower, -75)) {
					rc.transferPaint(nearestTower, -75);
				}
				return ;
			}
    		
    		if (weaving) {
				Direction dir = rc.getLocation().directionTo(myEnemyTower);
				if (moveAway) {
					if (rc.canAttack(myEnemyTower)){
						rc.setIndicatorString("ATTACKING1");
						rc.attack(myEnemyTower);
					}
					dir = dir.opposite();
					if (rc.canMove(dir)){
						rc.move(dir);
						rc.setIndicatorString("GOING AWAY");
					}

				}
				else {
					if (rc.canMove(dir)){
						rc.move(dir);
						rc.setIndicatorString("GOING TOWARDS" + myEnemyTower);
					}
					if (rc.canAttack(myEnemyTower)){
						rc.setIndicatorString("ATTACKING2");
						rc.attack(myEnemyTower);
					}
					// else if(rc.canMove(dir.opposite())){
					// 	rc.move(dir.opposite());
					// }
					if (rc.senseRobotAtLocation(myEnemyTower)==null){
						weaving = false;
					}

				}
				moveAway = !moveAway;
				return;
			}
	    	// Sense information about all visible nearby tiles.
		    MapInfo[] nearbyTiles = rc.senseNearbyMapInfos();
		    // Search for a nearby ruin to complete.

		    //If we can see a ruin, move towards it
		    MapInfo curRuin = null;
		    int ruinDist = 999999;
		    int towerDist = 999999;
		    for (MapInfo tile : nearbyTiles){
		    	RobotInfo potentialTower = rc.senseRobotAtLocation(tile.getMapLocation());
		    	if (tile.hasRuin() && potentialTower == null){
	            	int dist = tile.getMapLocation().distanceSquaredTo(rc.getLocation());
	            	if (dist < ruinDist) {
		                curRuin = tile;
		                ruinDist = dist;
	            	}
	            }
		    	else if (tile.hasRuin() && potentialTower != null) {
					if (potentialTower.getTeam() == rc.getTeam()){
						int dist = tile.getMapLocation().distanceSquaredTo(rc.getLocation());
						if (dist < towerDist) {
							nearestTower = tile.getMapLocation();
							towerDist = dist;
							if (rc.getPaint() < 100) {
								if (rc.canTransferPaint(nearestTower, -75)) {
									rc.transferPaint(nearestTower, -75);
								}
							}
						}
						if (rc.canUpgradeTower(tile.getMapLocation())) {
							rc.upgradeTower(tile.getMapLocation());
							System.out.println("Tower was upgraded!");
						}
					}
					else{
						//enemy tower!!
						Direction dir = rc.getLocation().directionTo(potentialTower.location);
						myEnemyTower = potentialTower.location;
						rc.setIndicatorString("updating myEnemyTower" + myEnemyTower);
						weaving = true;
						// if (!weaving && potentialTower.getType()!=UnitType.LEVEL_TWO_DEFENSE_TOWER && potentialTower.getType()!=UnitType.LEVEL_ONE_DEFENSE_TOWER && potentialTower.getType()!=UnitType.LEVEL_THREE_DEFENSE_TOWER) weaving=true;
						if (rc.canMove(dir)){
							rc.move(dir);
							moveAway = true;
						}
						if (rc.canAttack(myEnemyTower)){
							rc.attack(myEnemyTower);
						}
						return;
					}
	    		}
	    	}
		    
		    if (curRuin != null){
		    	paintingRuinLoc = curRuin.getMapLocation();
                paintingTowerType = getNewTowerType(rc);
                turnsWithoutAttack = 0;
                paintingTurns = 0;
                runPaintPattern(rc);
                return ;
	        }

	        if (rc.getMovementCooldownTurns() > 10) {
	        	Clock.yield();
	        }
	        if (true) {
	        	if ((rc.getLocation().x > mapWidth * 0.4 || rc.getLocation().x < mapWidth * 0.6)
		        		&& (rc.getLocation().y > mapHeight * 0.6 || rc.getLocation().y < mapHeight * 0.4)
		        		&& rc.getLocation().x % 4 == 0 && rc.getLocation().y % 4 == 3) {
		        	boolean has_mark = false;
		        	for (MapInfo patternTile : rc.senseNearbyMapInfos(rc.getLocation(), -1)){
		                if (patternTile.getMark() != patternTile.getPaint() && patternTile.getMark() != PaintType.EMPTY){
		                    boolean useSecondaryColor = patternTile.getMark() == PaintType.ALLY_SECONDARY;
		                    if (patternTile.getMark() != PaintType.EMPTY) {
		                    	has_mark = true;
		                    }
		                    if (rc.canAttack(patternTile.getMapLocation())) {
								rc.attack(patternTile.getMapLocation(), useSecondaryColor);
							}
		                }
		            }

		        	if (rc.getNumberTowers() > 5 && !has_mark && rc.canMarkResourcePattern(rc.getLocation())) {
		        		curResource = rc.getLocation();
		        		target = curResource;
		        		rc.markResourcePattern(curResource);
		        	}
		        }

		        if (curResource != null) {
		        	Direction dir = rc.getLocation().directionTo(curResource);

		            if (rc.getMovementCooldownTurns() > 10) {
		            	// Do nothing
		            }
		            else if (rc.canMove(dir)) {
		            	rc.move(dir);
		            }
		            else if (rc.canMove(dir.rotateRight())){
		            	rc.move(dir.rotateRight());
		            }
		            if (rc.canCompleteResourcePattern(curResource)) {
		            	rc.completeResourcePattern(curResource);
		            	curResource = null;
		            	break;
		            }
		        	for (MapInfo patternTile : rc.senseNearbyMapInfos(curResource, -1)){
		                if (patternTile.getMark() != patternTile.getPaint() && patternTile.getMark() != PaintType.EMPTY){
		                    boolean useSecondaryColor = patternTile.getMark() == PaintType.ALLY_SECONDARY;
		                    if (rc.canAttack(patternTile.getMapLocation())) {
								rc.attack(patternTile.getMapLocation(), useSecondaryColor);
							}
		                }
		            }
		        }
	        }

	        if (curRuin == null && curResource == null) {
	        	try {
	        		if (funcsplayer2.Util.distance(rc.getLocation(), target) <= 5 || !rc.isMovementReady()) {
	        			MapLocation oldTarget = target;
	        			while (Util.distance(funcsplayer2.Explore.getExploreTarget(), oldTarget) <= Math.max(mapHeight, mapWidth) / 2) {
	        				funcsplayer2.Explore.getNewTarget(10);
	        			}
	            		target = funcsplayer2.Explore.exploreTarget;
	        		}
	        		rc.setIndicatorString("Moving toward target at "+ target);
	        		funcsplayer2.Pathfinding.move(target, false);
	    		}
	        	catch (Exception e) {
	    			System.out.println(rc.getType() + " Exception");
	    		}
	        }
		    Clock.yield();
    	}
    }
    
    public static boolean isWithinPattern(MapLocation ruinLoc, MapLocation paintLoc) {
    	return Math.abs(paintLoc.x - ruinLoc.x) <= 2 && Math.abs(paintLoc.y - ruinLoc.y) <= 2 && !ruinLoc.equals(paintLoc);
    }
    
    public static boolean getIsSecondary(MapLocation ruinLoc, MapLocation paintLoc, UnitType towerType) {
        if(!isWithinPattern(ruinLoc, paintLoc)) return false;
        int col = paintLoc.x - ruinLoc.x + 2;
        int row = paintLoc.y - ruinLoc.y + 2;
        if (towerType == UnitType.LEVEL_ONE_PAINT_TOWER) {
        	return paintTowerPattern[row][col];
        }
        else if (towerType == UnitType.LEVEL_ONE_MONEY_TOWER) {
        	return moneyTowerPattern[row][col];
        }
        else {
        	return defenseTowerPattern[row][col];
        }
    }
    
    public static void runPaintPattern(RobotController rc) throws GameActionException {
    	/*
    	if(paintingTurns % 3 == 0) {
            Direction toRuin = rc.getLocation().directionTo(paintingRuinLoc);
            Direction tangent = toRuin.rotateRight().rotateRight();
            int distance = rc.getLocation().distanceSquaredTo(paintingRuinLoc);

            if(distance > 4) {
                //don't make the circle too big
                tangent = tangent.rotateLeft();
            }

            if(rc.canMove(tangent)) rc.move(tangent);
            else if(rc.canMove(tangent.rotateRight())) rc.move(tangent.rotateRight());
            else if(rc.canMove(tangent.rotateLeft())) rc.move(tangent.rotateLeft());
        }
        */
    	rc.setIndicatorString("SEARCHING FOR NEARBY ROBOTS");
    	for (RobotInfo robot: rc.senseNearbyRobots(-1)) {
	    	if (robot.team == rc.getTeam() && robot.type == UnitType.SOLDIER &&
	    			paintingRuinLoc.distanceSquaredTo(robot.location) < rc.getLocation().distanceSquaredTo(paintingRuinLoc)) {
	    		funcsplayer2.Pathfinding.move(target, false);
	    		return ;
	    	}
	    }
    	
        //paint tiles for the pattern
        if(rc.isActionReady()) {
            MapInfo[] infos = rc.senseNearbyMapInfos();
            boolean attacked = false;
            for(MapInfo info : infos) {
                MapLocation paintLoc = info.getMapLocation();
                boolean shouldBeSecondary = getIsSecondary(paintingRuinLoc, paintLoc, paintingTowerType);
                if(rc.canAttack(paintLoc)
                        && (info.getPaint() == PaintType.EMPTY || info.getPaint().isSecondary() != shouldBeSecondary) 
                        && isWithinPattern(paintingRuinLoc, paintLoc)) {
                    rc.attack(paintLoc, shouldBeSecondary);
                    attacked = true;
                    turnsWithoutAttack = 0;
                    break;
                }
            }
            if(!attacked) turnsWithoutAttack++;
        }
        
        Direction dir = rc.getLocation().directionTo(paintingRuinLoc);

        if (rc.getMovementCooldownTurns() > 10) {
        	// Do nothing
        }
        else if (rc.canMove(dir)) {
        	rc.move(dir);
        }
        else if (rc.canMove(dir.rotateRight())){
        	rc.move(dir.rotateRight());
        }

        if (rc.canCompleteTowerPattern(paintingTowerType, paintingRuinLoc)){
            rc.completeTowerPattern(paintingTowerType, paintingRuinLoc);
        }
    }
    
    public static UnitType getNewTowerType(RobotController rc) {
    	Random random = new Random();
        int randomNumber = random.nextInt(7);

        boolean inMiddleX = 0.3 * mapWidth < paintingRuinLoc.x && paintingRuinLoc.x < 0.7 * mapWidth;
        boolean inMiddleY = 0.3 * mapHeight < paintingRuinLoc.y && paintingRuinLoc.y < 0.7 * mapHeight;
        if (inMiddleX && inMiddleY) {
    		return UnitType.LEVEL_ONE_DEFENSE_TOWER;
    	}
        else if (rc.getNumberTowers() < 4) {
        	return UnitType.LEVEL_ONE_MONEY_TOWER;
        }
    	else if (randomNumber == 0 || randomNumber == 1) {
	    	return UnitType.LEVEL_ONE_PAINT_TOWER;
    	}
	    else {
	    	return UnitType.LEVEL_ONE_MONEY_TOWER;
	    }
    }
}